File: Decorator/Java/src/StarbuzzCoffee.java
package eu.svetlin.design.patterns.decorator;

import eu.svetlin.design.patterns.decorator.beverages.Beverage;
import eu.svetlin.design.patterns.decorator.beverages.DarkRoast;
import eu.svetlin.design.patterns.decorator.beverages.Espresso;
import eu.svetlin.design.patterns.decorator.beverages.HouseBlend;
import eu.svetlin.design.patterns.decorator.condiments.Mocha;
import eu.svetlin.design.patterns.decorator.condiments.Soy;
import eu.svetlin.design.patterns.decorator.condiments.Whip;

public class StarbuzzCoffee {
    public static void main(String args[]){
        Beverage espresso = new Espresso();
        System.out.println(espresso.getDescription() + " $" + espresso.cost());

        Beverage doubleMochaWithWhip = new DarkRoast();
        doubleMochaWithWhip = new Mocha(doubleMochaWithWhip);
        doubleMochaWithWhip = new Mocha(doubleMochaWithWhip);
        doubleMochaWithWhip = new Whip(doubleMochaWithWhip);

        System.out.println(doubleMochaWithWhip.getDescription() + " $" + doubleMochaWithWhip.cost());

        Beverage houseBlendWithSoyMochaAndWhip = new HouseBlend();
        houseBlendWithSoyMochaAndWhip.setSize(Beverage.Size.VENTI);
        houseBlendWithSoyMochaAndWhip = new Soy(houseBlendWithSoyMochaAndWhip);
        houseBlendWithSoyMochaAndWhip = new Mocha(houseBlendWithSoyMochaAndWhip);
        houseBlendWithSoyMochaAndWhip = new Whip(houseBlendWithSoyMochaAndWhip);

        System.out.println(houseBlendWithSoyMochaAndWhip.getDescription() + " $" + String.format("%.2f", houseBlendWithSoyMochaAndWhip.cost()));
    }
}

File: Decorator/TypeScript/src/StarBuzzCoffee.ts
import Beverage from './beverages/Beverage';
import DarkRoast from './beverages/DarkRoast';
import Espresso from './beverages/Espresso';
import HouseBlend from './beverages/HouseBlend';
import Mocha from './condiments/Mocha';
import Whip from './condiments/Whip';
import Soy from './condiments/Soy';

const main = () => {
  const espresso: Beverage = new Espresso();
  console.log(espresso.getDescription(), ' $' + espresso.cost());

  let doubleMochaWithWhip = new DarkRoast();
  doubleMochaWithWhip = new Mocha(doubleMochaWithWhip);
  doubleMochaWithWhip = new Mocha(doubleMochaWithWhip);
  doubleMochaWithWhip = new Whip(doubleMochaWithWhip);

  console.log(
    doubleMochaWithWhip.getDescription(),
    ' $' + doubleMochaWithWhip.cost()
  );

  let houseBlendWithSoyMochaAndWhip = new HouseBlend();
  houseBlendWithSoyMochaAndWhip = new Soy(houseBlendWithSoyMochaAndWhip);
  houseBlendWithSoyMochaAndWhip = new Mocha(houseBlendWithSoyMochaAndWhip);
  houseBlendWithSoyMochaAndWhip = new Whip(houseBlendWithSoyMochaAndWhip);

  console.log(
    houseBlendWithSoyMochaAndWhip.getDescription(),
    ' $' + houseBlendWithSoyMochaAndWhip.cost()
  );
};

main();
